modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java

```diff
@@ -162,6 +175,142 @@ public void testConsistencyAfterResettingLostPartitions_3() throws Exception {
+     */
+    @Test
+    public void testConsistencyAfterClusterRestart() throws Exception {
+        doTestConsistencyAfterClusterRestartWithDisabledWalOnRebalancing(true);
+    }
+
+    /**
+     * Tests data recovery after restarting the cluster with disabled WAL on rebalancing.
+     *
+     * @throws Exception If failed.
+     */
+    public void doTestConsistencyAfterClusterRestartWithDisabledWalOnRebalancing(boolean stopInTheMiddleOfCheckpoint) throws Exception {
+        lossPlc = READ_WRITE_SAFE;
+
+        IgniteEx ig0 = startGrids(2);
+
+        ig0.cluster().state(ACTIVE);
+
+        IgniteCache<Integer, Integer> cache = ig0.cache(DEFAULT_CACHE_NAME);
+
+        // Initial load.
+        for (int k = 0; k < 2500; k++)
+            cache.put(k, k);
+
+        // Stop the second node and update data in the cache to trigger rebalancing when the node will be restarted.
+        stopGrid(1);
+
+        int owningPartId = 12;
+
+        // Rewrite data to trigger further rebalance.
+        for (int k = 2500; k < 5000; k++) {
+            if (!stopInTheMiddleOfCheckpoint) {
+                // Update all partitions in oder to disable WAL on rebalancing.
+                // It is possible when there are no partitions in OWNING state.
+                cache.put(k, k);
+            }
+            else {
+                // Should skip one random partition to be sure that after restarting the second node,
+                // it will have at least one partition in OWNING state, and so WAL will not be disabled while rebalancing.
+                if (ig0.affinity(DEFAULT_CACHE_NAME).partition(k) != owningPartId)
+                    cache.put(k, k);
+            }
+        }
+
+        // Block supply messages to prevent partitions from being moved to the second node.
+        TestRecordingCommunicationSpi spi0 = (TestRecordingCommunicationSpi) ig0.configuration().getCommunicationSpi();
+        spi0.blockMessages(TestRecordingCommunicationSpi.blockSupplyMessageForGroup(CU.cacheId(DEFAULT_CACHE_NAME)));
+
+        // Restart the second node.
+        IgniteEx ig1 = startGrid(1);
+
+        // Wait for exchange. All partitions should be moved to the MOVING state.
+        ig1.context().cache().context().exchange().lastTopologyFuture().get();
+
+        // Stopping the coordinator node leads to moving all partitions to the LOST state.
+        stopGrid(0);
+
+        for (int partId = 0; partId < PARTS_CNT; partId++) {
+            if (stopInTheMiddleOfCheckpoint && partId == owningPartId)
+                continue;
+
+            checkLostPartitionAcrossCluster(DEFAULT_CACHE_NAME, partId);
+        }
+
+        assertEquals(
+            "WAL should be " + (stopInTheMiddleOfCheckpoint ? "enabled" : "disabled") + " for the cache + " + DEFAULT_CACHE_NAME,
+            stopInTheMiddleOfCheckpoint,
+            ig1.context().cache().cacheGroup(CU.cacheId(DEFAULT_CACHE_NAME)).localWalEnabled());
+
+        if (!stopInTheMiddleOfCheckpoint)
```

Comment:
I would not use a negative over a whole condition in a clause with else.

Reply:
Fixed